tests/rust/safe_abstraction/cell_u32.rs

```diff
@@ -92,5 +100,33 @@ impl CellU32 {
         //@ close CellU32_own(t, u);
         //@ close CellU32_nonatomic_borrow_content(self, t)();
         //@ close_nonatomic_borrow();
+        //@ thread_token_merge(t, MaskTopMinus([pair(self, MaskTopMinus([]))]), m);
+    }
+
+    fn swap<'a>(&'a self, other: &'a Self) {
+        //@ open CellU32_share(a, _t, self);
+        //@ open CellU32_share(a, _t, other);
+        if self as *const CellU32 == other as *const CellU32 {
+            return;
+        }
+        //@ assert [_]nonatomic_borrow(a, _t, ?ms, CellU32_nonatomic_borrow_content(self, _t));
+        //@ assert [_]nonatomic_borrow(a, _t, ?mo, CellU32_nonatomic_borrow_content(other, _t));
+        //@ thread_token_split(_t, MaskTopMinus([]), ms);
+        //@ thread_token_split(_t, MaskTopMinus([pair(self, MaskTopMinus([]))]), mo);
+        //@ open_nonatomic_borrow(a, _t, ms, _q_a/2);
+        //@ open_nonatomic_borrow(a, _t, mo, _q_a/2);
+        let ps = self.v.get();
+        let po = other.v.get();
+        unsafe {
+            let tmp = *po;
+            *po = *ps;
+            *ps = tmp;
+        }
+        //@ assert thread_token(_t, ?rem);
+        //@ close_nonatomic_borrow();
+        //@ close_nonatomic_borrow();
+        //@ thread_token_merge(_t, rem, mo);
+        //@ assert thread_token(_t, ms) &*& thread_token(_t, ?rem1);
+        //@ thread_token_merge(_t, rem1, ms);
```

Comment:
Very nice! The fact that this proof goes through shows that mask.gh is fine computationally.